From 179ff511fcd222fb5738d187770c5bdb83d03dae Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Fri, 15 Mar 2024 15:38:12 +0100
Subject: [PATCH 0177/1013] igb: add oob I/O capabilities

Enable the IGB driver for out-of-band I/O traffic. In a nutshell, this
entails the following steps:

- turn on/off oob delivery for the NIC interrupts upon request from
  the in-band stack.

- use an oob-accessible page pool for RX.

- make sure TX only synchronizes the CPU caches instead of streaming
  DMA buffers when receiving data living in an oob-accessible page
  pool.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 drivers/net/ethernet/intel/Kconfig           |   9 +
 drivers/net/ethernet/intel/igb/igb.h         |  63 +-
 drivers/net/ethernet/intel/igb/igb_ethtool.c |   8 +-
 drivers/net/ethernet/intel/igb/igb_main.c    | 625 +++++++++++++++----
 drivers/net/ethernet/intel/igb/igb_ptp.c     |  58 +-
 5 files changed, 616 insertions(+), 147 deletions(-)

diff --git a/drivers/net/ethernet/intel/Kconfig b/drivers/net/ethernet/intel/Kconfig
index 0375c7448a57..6e27c8f4ff4b 100644
--- a/drivers/net/ethernet/intel/Kconfig
+++ b/drivers/net/ethernet/intel/Kconfig
@@ -106,6 +106,15 @@ config IGB
 	  To compile this driver as a module, choose M here. The module
 	  will be called igb.
 
+config IGB_OOB
+	bool "Out-of-band support for Intel(R) 82575/82576 Gigabit Ethernet"
+	depends on NET_OOB
+	help
+	  Enable out-of-band I/O. Say Y if a companion kernel may divert
+	  the network traffic going through IGB devices for real-time delivery.
+	  Enabling this option makes sense only in a dual kernel configuration.
+	  If unsure, disable it.
+
 config IGB_HWMON
 	bool "Intel(R) PCI-Express Gigabit adapters HWMON support"
 	default y
diff --git a/drivers/net/ethernet/intel/igb/igb.h b/drivers/net/ethernet/intel/igb/igb.h
index 3c2dc7bdebb5..d5af48090abe 100644
--- a/drivers/net/ethernet/intel/igb/igb.h
+++ b/drivers/net/ethernet/intel/igb/igb.h
@@ -18,6 +18,7 @@
 #include <linux/i2c-algo-bit.h>
 #include <linux/pci.h>
 #include <linux/mdio.h>
+#include <linux/irq_work.h>
 
 #include <net/xdp.h>
 
@@ -25,6 +26,17 @@ struct igb_adapter;
 
 #define E1000_PCS_CFG_IGN_SD	1
 
+/* The ICR conditions we should process in-band via interrupt
+ * forwarding between execution stages (i.e. detected oob but handled
+ * in-band). */
+#define E1000_ICR_FORWARD_MASK					\
+	(E1000_ICR_DRSTA|E1000_ICR_DOUTSYNC|			\
+		E1000_ICR_RXSEQ|E1000_ICR_LSC|E1000_ICR_TS)
+
+#define E1000_ICR_OTHER_FORWARD_MASK				\
+	(E1000_ICR_DRSTA|E1000_ICR_DOUTSYNC|			\
+		E1000_ICR_VMMB|E1000_ICR_LSC|E1000_ICR_TS)
+
 /* Interrupt defines */
 #define IGB_START_ITR		648 /* ~6000 ints/sec */
 #define IGB_4K_ITR		980
@@ -225,6 +237,9 @@ enum igb_tx_flags {
 	/* olinfo flags */
 	IGB_TX_FLAGS_IPV4	= 0x10,
 	IGB_TX_FLAGS_CSUM	= 0x20,
+
+	/* oob management flags */
+	IGB_TX_OOB		= 0x40,
 };
 
 /* VLAN info */
@@ -288,6 +303,18 @@ struct igb_rx_buffer {
 	__u16 pagecnt_bias;
 };
 
+struct igb_inband_work {
+	struct igb_ring *ring;
+	dma_addr_t dma;
+	union {
+		struct {
+			struct page *page;
+			__u16 pagecnt_bias;
+		};
+		unsigned int len;
+	};
+};
+
 struct igb_tx_queue_stats {
 	u64 packets;
 	u64 bytes;
@@ -326,6 +353,11 @@ struct igb_ring {
 	void __iomem *tail;		/* pointer to ring tail register */
 	dma_addr_t dma;			/* phys address of the ring */
 	unsigned int  size;		/* length of desc. ring in bytes */
+#ifdef CONFIG_IGB_OOB
+	struct page_pool *rx_oob_pool;
+	struct igb_inband_work *inband_flush;
+	struct irq_work inband_irq_work;
+#endif
 
 	u16 count;			/* number of desc. in the ring */
 	u8 queue_index;			/* logical index of the ring*/
@@ -341,6 +373,10 @@ struct igb_ring {
 	u16 next_to_clean;
 	u16 next_to_use;
 	u16 next_to_alloc;
+#ifdef CONFIG_IGB_OOB
+	u16 next_to_defer;
+	u16 next_to_flush;
+#endif
 
 	union {
 		/* TX */
@@ -560,6 +596,11 @@ struct igb_adapter {
 	int num_rx_queues;
 	struct igb_ring *rx_ring[16];
 
+#ifdef CONFIG_IGB_OOB
+	u32 cached_icr;
+	u32 cached_tsicr;
+#endif
+
 	u32 max_frame_size;
 	u32 min_frame_size;
 
@@ -623,7 +664,7 @@ struct igb_adapter {
 	unsigned long last_rx_ptp_check;
 	unsigned long last_rx_timestamp;
 	unsigned int ptp_flags;
-	spinlock_t tmreg_lock;
+	hard_spinlock_t tmreg_lock;
 	struct cyclecounter cc;
 	struct timecounter tc;
 	u32 tx_hwtstamp_timeouts;
@@ -797,6 +838,26 @@ static inline struct netdev_queue *txring_txq(const struct igb_ring *tx_ring)
 	return netdev_get_tx_queue(tx_ring->netdev, tx_ring->queue_index);
 }
 
+/*
+ * Check whether oob support is compiled in for the IGB driver. This
+ * tells nothing about the current execution stage, or whether oob
+ * diversion is ongoing for any IGB device.
+ */
+static inline bool igb_net_oob(void)
+{
+	return IS_ENABLED(CONFIG_IGB_OOB);
+}
+
+/*
+ * Check whether the caller is running on the out-of-band stage, with
+ * the precondition that oob support is compiled in for the IGB
+ * driver.
+ */
+static inline bool igb_running_oob(void)
+{
+	return igb_net_oob() && running_oob();
+}
+
 int igb_add_filter(struct igb_adapter *adapter,
 		   struct igb_nfc_filter *input);
 int igb_erase_filter(struct igb_adapter *adapter,
diff --git a/drivers/net/ethernet/intel/igb/igb_ethtool.c b/drivers/net/ethernet/intel/igb/igb_ethtool.c
index 6412c84e2d17..4d12b8f30078 100644
--- a/drivers/net/ethernet/intel/igb/igb_ethtool.c
+++ b/drivers/net/ethernet/intel/igb/igb_ethtool.c
@@ -1920,7 +1920,9 @@ static int igb_run_loopback_test(struct igb_adapter *adapter)
 		/* place 64 packets on the transmit queue*/
 		for (i = 0; i < 64; i++) {
 			skb_get(skb);
+			netif_tx_lock_oob(txring_txq(tx_ring));
 			tx_ret_val = igb_xmit_frame_ring(skb, tx_ring);
+			netif_tx_unlock_oob(txring_txq(tx_ring));
 			if (tx_ret_val == NETDEV_TX_OK)
 				good_cnt++;
 		}
@@ -3433,8 +3435,12 @@ static int igb_set_priv_flags(struct net_device *netdev, u32 priv_flags)
 	unsigned int flags = adapter->flags;
 
 	flags &= ~IGB_FLAG_RX_LEGACY;
-	if (priv_flags & IGB_PRIV_FLAGS_LEGACY_RX)
+	if (priv_flags & IGB_PRIV_FLAGS_LEGACY_RX) {
+		/* Legacy RX is unavailable when oob mode is on. */
+		if (netif_oob_diversion(netdev))
+			return -EPERM;
 		flags |= IGB_FLAG_RX_LEGACY;
+	}
 
 	if (flags != adapter->flags) {
 		adapter->flags = flags;
diff --git a/drivers/net/ethernet/intel/igb/igb_main.c b/drivers/net/ethernet/intel/igb/igb_main.c
index 18284a838e24..e280c24d46d6 100644
--- a/drivers/net/ethernet/intel/igb/igb_main.c
+++ b/drivers/net/ethernet/intel/igb/igb_main.c
@@ -6,6 +6,7 @@
 #include <linux/module.h>
 #include <linux/types.h>
 #include <linux/init.h>
+#include <linux/irq.h>
 #include <linux/bitops.h>
 #include <linux/vmalloc.h>
 #include <linux/pagemap.h>
@@ -16,6 +17,8 @@
 #include <net/ip6_checksum.h>
 #include <net/pkt_sched.h>
 #include <net/pkt_cls.h>
+#include <net/page_pool/helpers.h>
+#include <linux/skbuff_ref.h>
 #include <linux/net_tstamp.h>
 #include <linux/mii.h>
 #include <linux/ethtool.h>
@@ -726,6 +729,225 @@ u32 igb_rd32(struct e1000_hw *hw, u32 reg)
 	return value;
 }
 
+static void __igb_flush_page(struct igb_ring *rx_ring, dma_addr_t dma,
+			struct page *page, __u16 bias)
+{
+	dma_unmap_page_attrs(rx_ring->dev, dma,
+			igb_rx_pg_size(rx_ring), DMA_FROM_DEVICE,
+			IGB_RX_DMA_ATTR);
+	__page_frag_cache_drain(page, bias);
+}
+
+#ifdef CONFIG_IGB_OOB
+
+static int igb_enable_oob(struct net_device *netdev)
+{
+	struct igb_adapter *adapter = netdev_priv(netdev);
+	int ret = 0, n;
+
+	/* Catch late init errors for oob mode, still allowing
+	 * operations in regular mode though. */
+	for (n = 0; n < adapter->num_rx_queues; n++)
+		if (!adapter->rx_ring[n]->rx_oob_pool)
+			return -ENOMEM;
+
+	if (adapter->flags & IGB_FLAG_HAS_MSIX) {
+		pr_info("%s: enabling out-of-band I/O mode (MSI-X)\n", netdev_name(netdev));
+		for (n = 0; n <= adapter->num_q_vectors && n <= MAX_Q_VECTORS; n++) {
+			ret = irq_switch_oob(adapter->msix_entries[n].vector, true);
+			if (ret) {
+				while (--n >= 0)
+					irq_switch_oob(adapter->msix_entries[n].vector, false);
+				break;
+			}
+		}
+	} else {
+		ret = irq_switch_oob(adapter->pdev->irq, true);
+		pr_info("%s: enabling out-of-band I/O mode (%s)\n",
+			netdev_name(netdev),
+			adapter->flags & IGB_FLAG_HAS_MSI ? "MSI" : "legacy-int");
+	}
+
+	return ret;
+}
+
+static void igb_disable_oob(struct net_device *netdev)
+{
+	struct igb_adapter *adapter = netdev_priv(netdev);
+	int n;
+
+	if (adapter->flags & IGB_FLAG_HAS_MSIX) {
+		pr_info("%s: disabling out-of-band I/O mode (MSI-X)\n", netdev_name(netdev));
+		for (n = 0; n <= adapter->num_q_vectors && n <= MAX_Q_VECTORS; n++)
+			irq_switch_oob(adapter->msix_entries[n].vector, false);
+	} else {
+		irq_switch_oob(adapter->pdev->irq, false);
+		pr_info("%s: disabling out-of-band I/O mode (%s)\n",
+			netdev_name(netdev),
+			adapter->flags & IGB_FLAG_HAS_MSI ? "MSI" : "legacy-int");
+	}
+}
+
+static void igb_create_oob_pool(struct igb_ring *ring)
+{
+	struct igb_adapter *adapter = netdev_priv(ring->netdev);
+	struct page_pool_params params = {
+		.order = igb_rx_pg_order(ring),
+		.flags = PP_FLAG_DMA_MAP | PP_FLAG_PAGE_OOB,
+		.pool_size = ring->count * 4,
+		.nid = dev_to_node(&adapter->pdev->dev),
+		.dev = &adapter->pdev->dev,
+		.dma_dir = DMA_FROM_DEVICE,
+		.offset = 0,
+		.max_len = igb_rx_pg_size(ring),
+	};
+	struct page_pool *rx_oob_pool = page_pool_create(&params);
+
+	/*
+	 * On error allocating the oob pool, we can still operate in
+	 * regular in-band mode, rejecting oob traffic though.
+	 */
+	if (WARN_ON(IS_ERR(rx_oob_pool)))
+		return;
+
+	ring->rx_oob_pool = rx_oob_pool;
+}
+
+static inline bool igb_is_oob_page(struct igb_ring *rx_ring,
+				struct page *page)
+{
+	return napi_pp_get_pool(page_to_netmem(page)) == rx_ring->rx_oob_pool;
+}
+
+static void igb_rx_inband_work(struct irq_work *irq_work)
+{
+	struct igb_ring *ring = container_of(irq_work, struct igb_ring, inband_irq_work);
+
+	while (ring->next_to_flush != READ_ONCE(ring->next_to_defer)) {
+		u16 ntf = ring->next_to_flush;
+		struct igb_inband_work *fl = ring->inband_flush + ntf;
+		__igb_flush_page(ring, fl->dma, fl->page, fl->pagecnt_bias);
+		ring->next_to_flush = (ntf + 1) % ring->count;
+	}
+}
+
+static void igb_flush_page(struct igb_ring *rx_ring,
+			struct igb_rx_buffer *rx_buffer)
+{
+	/* If running oob, defer the release to the inband stage. We
+	 * get there only for pages which do NOT belong to the oob
+	 * pool.
+	 */
+	if (igb_running_oob()) {
+		u16 ntd = rx_ring->next_to_defer;
+		struct igb_inband_work *fl = rx_ring->inband_flush + ntd;
+		/* We are called by the napi ->poll() handler, which
+		 * is serialized in oob context too. */
+		fl->page = rx_buffer->page;
+		fl->pagecnt_bias = rx_buffer->pagecnt_bias;
+		fl->dma = rx_buffer->dma;
+		fl->ring = rx_ring;
+		WRITE_ONCE(rx_ring->next_to_defer, (ntd + 1) % rx_ring->count);
+		irq_work_queue(&rx_ring->inband_irq_work);
+	} else {
+		__igb_flush_page(rx_ring, rx_buffer->dma, rx_buffer->page,
+				rx_buffer->pagecnt_bias);
+	}
+}
+
+static void igb_tx_inband_work(struct irq_work *irq_work)
+{
+	struct igb_ring *ring = container_of(irq_work, struct igb_ring, inband_irq_work);
+
+	while (ring->next_to_flush != READ_ONCE(ring->next_to_defer)) {
+		u16 ntf = ring->next_to_flush;
+		struct igb_inband_work *fl = ring->inband_flush + ntf;
+		dma_unmap_single(fl->ring->dev,	fl->dma, fl->len, DMA_TO_DEVICE);
+		ring->next_to_flush = (ntf + 1) % ring->count;
+	}
+}
+
+static void igb_unmap_buffer(struct igb_ring *tx_ring,
+			struct igb_tx_buffer *tx_buffer)
+{
+	if (igb_running_oob()) {
+		u16 ntd = tx_ring->next_to_defer;
+		struct igb_inband_work *fl = tx_ring->inband_flush + ntd;
+		fl->dma = dma_unmap_addr(tx_buffer, dma);
+		fl->len = dma_unmap_len(tx_buffer, len);
+		fl->ring = tx_ring;
+		WRITE_ONCE(tx_ring->next_to_defer, (ntd + 1) % tx_ring->count);
+		irq_work_queue(&tx_ring->inband_irq_work);
+	} else {
+		dma_unmap_single(tx_ring->dev,
+				dma_unmap_addr(tx_buffer, dma),
+				dma_unmap_len(tx_buffer, len),
+				DMA_TO_DEVICE);
+	}
+}
+
+static inline u32 igb_get_icr(struct igb_adapter *adapter)
+{
+	struct e1000_hw *hw = &adapter->hw;
+	u32 icr;
+
+	if (igb_net_oob() && running_inband()) {
+		icr = adapter->cached_icr;
+		adapter->cached_icr = 0;
+		return icr;
+	}
+
+	adapter->cached_icr |= rd32(E1000_ICR);
+
+	return adapter->cached_icr;
+}
+
+static inline void igb_clear_icr(struct igb_adapter *adapter)
+{
+	adapter->cached_icr = 0;
+}
+
+#else  /* !CONFIG_IGB_OOB */
+
+static inline void igb_create_oob_pool(struct igb_ring *ring)
+{
+}
+
+static inline bool igb_is_oob_page(struct igb_ring *rx_ring,
+				struct page *page)
+{
+	return false;
+}
+
+static void igb_flush_page(struct igb_ring *rx_ring,
+			struct igb_rx_buffer *rx_buffer)
+{
+	__igb_flush_page(rx_ring, rx_buffer->dma, rx_buffer->page,
+			rx_buffer->pagecnt_bias);
+}
+
+static void igb_unmap_buffer(struct igb_ring *tx_ring,
+			struct igb_tx_buffer *tx_buffer)
+{
+	dma_unmap_single(tx_ring->dev,
+			dma_unmap_addr(tx_buffer, dma),
+			dma_unmap_len(tx_buffer, len),
+			DMA_TO_DEVICE);
+}
+
+static inline u32 igb_get_icr(struct igb_adapter *adapter)
+{
+	struct e1000_hw *hw = &adapter->hw;
+
+	return rd32(E1000_ICR);
+}
+
+static inline void igb_clear_icr(struct igb_adapter *adapter)
+{
+}
+
+#endif	/* !CONFIG_IGB_OOB */
+
 /**
  *  igb_write_ivar - configure ivar for given MSI-X vector
  *  @hw: pointer to the HW structure
@@ -1990,6 +2212,7 @@ static void igb_configure(struct igb_adapter *adapter)
 	 */
 	for (i = 0; i < adapter->num_rx_queues; i++) {
 		struct igb_ring *ring = adapter->rx_ring[i];
+		igb_create_oob_pool(ring);
 		igb_alloc_rx_buffers(ring, igb_desc_unused(ring));
 	}
 }
@@ -3042,6 +3265,10 @@ static const struct net_device_ops igb_netdev_ops = {
 	.ndo_setup_tc		= igb_setup_tc,
 	.ndo_bpf		= igb_xdp,
 	.ndo_xdp_xmit		= igb_xdp_xmit,
+#ifdef CONFIG_IGB_OOB
+	.ndo_enable_oob		= igb_enable_oob,
+	.ndo_disable_oob	= igb_disable_oob,
+#endif
 };
 
 /**
@@ -3629,6 +3856,11 @@ static int igb_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 		}
 	}
 
+	if (IS_ENABLED(CONFIG_IGB_OOB)) {
+		netdev_set_oob_capable(netdev);
+		netdev_info(netdev, "IGB device is oob-capable\n");
+	}
+
 	dev_pm_set_driver_flags(&pdev->dev, DPM_FLAG_NO_DIRECT_COMPLETE);
 
 	pm_runtime_put_noidle(&pdev->dev);
@@ -4276,6 +4508,18 @@ int igb_setup_tx_resources(struct igb_ring *tx_ring)
 	if (!tx_ring->tx_buffer_info)
 		goto err;
 
+#ifdef CONFIG_IGB_OOB
+	size = sizeof(struct igb_inband_work) * tx_ring->count;
+
+	tx_ring->inband_flush = vmalloc(size);
+	if (!tx_ring->inband_flush)
+		goto err;
+
+	init_irq_work(&tx_ring->inband_irq_work, igb_tx_inband_work);
+	tx_ring->next_to_defer = 0;
+	tx_ring->next_to_flush = 0;
+#endif
+
 	/* round up to nearest 4K */
 	tx_ring->size = tx_ring->count * sizeof(union e1000_adv_tx_desc);
 	tx_ring->size = ALIGN(tx_ring->size, 4096);
@@ -4293,6 +4537,10 @@ int igb_setup_tx_resources(struct igb_ring *tx_ring)
 err:
 	vfree(tx_ring->tx_buffer_info);
 	tx_ring->tx_buffer_info = NULL;
+#ifdef CONFIG_IGB_OOB
+	vfree(tx_ring->inband_flush);
+	tx_ring->inband_flush = NULL;
+#endif
 	dev_err(dev, "Unable to allocate memory for the Tx descriptor ring\n");
 	return -ENOMEM;
 }
@@ -4437,6 +4685,18 @@ int igb_setup_rx_resources(struct igb_ring *rx_ring)
 	if (!rx_ring->rx_buffer_info)
 		goto err;
 
+#ifdef CONFIG_IGB_OOB
+	size = sizeof(struct igb_inband_work) * rx_ring->count;
+
+	rx_ring->inband_flush = vmalloc(size);
+	if (!rx_ring->inband_flush)
+		goto err;
+
+	init_irq_work(&rx_ring->inband_irq_work, igb_rx_inband_work);
+	rx_ring->next_to_defer = 0;
+	rx_ring->next_to_flush = 0;
+#endif
+
 	/* Round up to nearest 4K */
 	rx_ring->size = rx_ring->count * sizeof(union e1000_adv_rx_desc);
 	rx_ring->size = ALIGN(rx_ring->size, 4096);
@@ -4458,6 +4718,10 @@ int igb_setup_rx_resources(struct igb_ring *rx_ring)
 	xdp_rxq_info_unreg(&rx_ring->xdp_rxq);
 	vfree(rx_ring->rx_buffer_info);
 	rx_ring->rx_buffer_info = NULL;
+#ifdef CONFIG_IGB_OOB
+	vfree(rx_ring->inband_flush);
+	rx_ring->inband_flush = NULL;
+#endif
 	dev_err(dev, "Unable to allocate memory for the Rx descriptor ring\n");
 	return -ENOMEM;
 }
@@ -4858,6 +5122,10 @@ void igb_free_tx_resources(struct igb_ring *tx_ring)
 
 	vfree(tx_ring->tx_buffer_info);
 	tx_ring->tx_buffer_info = NULL;
+#ifdef CONFIG_IGB_OOB
+	vfree(tx_ring->inband_flush);
+	tx_ring->inband_flush = NULL;
+#endif
 
 	/* if not set, then don't free */
 	if (!tx_ring->desc)
@@ -4884,6 +5152,11 @@ static void igb_free_all_tx_resources(struct igb_adapter *adapter)
 			igb_free_tx_resources(adapter->tx_ring[i]);
 }
 
+static inline bool igb_is_oob_tx(struct igb_tx_buffer *tx_buffer)
+{
+	return igb_net_oob() && tx_buffer->tx_flags & IGB_TX_OOB;
+}
+
 /**
  *  igb_clean_tx_ring - Free Tx Buffers
  *  @tx_ring: ring to be cleaned
@@ -4902,11 +5175,12 @@ static void igb_clean_tx_ring(struct igb_ring *tx_ring)
 		else
 			xdp_return_frame(tx_buffer->xdpf);
 
-		/* unmap skb header data */
-		dma_unmap_single(tx_ring->dev,
-				 dma_unmap_addr(tx_buffer, dma),
-				 dma_unmap_len(tx_buffer, len),
-				 DMA_TO_DEVICE);
+		/* unmap skb header data unless oob-managed */
+		if (!igb_is_oob_tx(tx_buffer))
+			dma_unmap_single(tx_ring->dev,
+					dma_unmap_addr(tx_buffer, dma),
+					dma_unmap_len(tx_buffer, len),
+					DMA_TO_DEVICE);
 
 		/* check for eop_desc to determine the end of the packet */
 		eop_desc = tx_buffer->next_to_watch;
@@ -4924,7 +5198,7 @@ static void igb_clean_tx_ring(struct igb_ring *tx_ring)
 			}
 
 			/* unmap any remaining paged data */
-			if (dma_unmap_len(tx_buffer, len))
+			if (dma_unmap_len(tx_buffer, len) && !igb_is_oob_tx(tx_buffer))
 				dma_unmap_page(tx_ring->dev,
 					       dma_unmap_addr(tx_buffer, dma),
 					       dma_unmap_len(tx_buffer, len),
@@ -4977,6 +5251,10 @@ void igb_free_rx_resources(struct igb_ring *rx_ring)
 	xdp_rxq_info_unreg(&rx_ring->xdp_rxq);
 	vfree(rx_ring->rx_buffer_info);
 	rx_ring->rx_buffer_info = NULL;
+#ifdef CONFIG_IGB_OOB
+	vfree(rx_ring->inband_flush);
+	rx_ring->inband_flush = NULL;
+#endif
 
 	/* if not set, then don't free */
 	if (!rx_ring->desc)
@@ -5017,6 +5295,7 @@ static void igb_clean_rx_ring(struct igb_ring *rx_ring)
 	/* Free all the Rx ring sk_buffs */
 	while (i != rx_ring->next_to_alloc) {
 		struct igb_rx_buffer *buffer_info = &rx_ring->rx_buffer_info[i];
+		struct page *page = buffer_info->page;
 
 		/* Invalidate cache lines that may have been written to by
 		 * device so that we avoid corrupting memory.
@@ -5027,14 +5306,19 @@ static void igb_clean_rx_ring(struct igb_ring *rx_ring)
 					      igb_rx_bufsz(rx_ring),
 					      DMA_FROM_DEVICE);
 
-		/* free resources associated with mapping */
-		dma_unmap_page_attrs(rx_ring->dev,
-				     buffer_info->dma,
-				     igb_rx_pg_size(rx_ring),
-				     DMA_FROM_DEVICE,
-				     IGB_RX_DMA_ATTR);
-		__page_frag_cache_drain(buffer_info->page,
-					buffer_info->pagecnt_bias);
+		/* We may find pages coming from an oob pool. */
+		if (igb_is_oob_page(rx_ring, page)) {
+			napi_pp_put_page(page_to_netmem(page));
+		} else {
+			/* free resources associated with mapping */
+			dma_unmap_page_attrs(rx_ring->dev,
+					buffer_info->dma,
+					igb_rx_pg_size(rx_ring),
+					DMA_FROM_DEVICE,
+					IGB_RX_DMA_ATTR);
+			__page_frag_cache_drain(page,
+						buffer_info->pagecnt_bias);
+		}
 
 		i++;
 		if (i == rx_ring->count)
@@ -5044,6 +5328,13 @@ static void igb_clean_rx_ring(struct igb_ring *rx_ring)
 	rx_ring->next_to_alloc = 0;
 	rx_ring->next_to_clean = 0;
 	rx_ring->next_to_use = 0;
+
+#ifdef CONFIG_IGB_OOB
+	if (rx_ring->rx_oob_pool) {
+		page_pool_destroy(rx_ring->rx_oob_pool);
+		rx_ring->rx_oob_pool = NULL;
+	}
+#endif
 }
 
 /**
@@ -5981,6 +6272,9 @@ static int igb_tso(struct igb_ring *tx_ring,
 	u32 paylen, l4_offset;
 	int err;
 
+	if (igb_running_oob())	/* no TSO for oob traffic. */
+		return 0;
+
 	if (skb->ip_summed != CHECKSUM_PARTIAL)
 		return 0;
 
@@ -6185,6 +6479,13 @@ static inline int igb_maybe_stop_tx(struct igb_ring *tx_ring, const u16 size)
 {
 	if (igb_desc_unused(tx_ring) >= size)
 		return 0;
+	/*
+	 * Dovetail: callers are always serialized, by some inband
+	 * stack lock, OR a mutual exclusion scheme between oob and
+	 * inband issuers enforced by the companion kernel (e.g. evl
+	 * stax), OR single-threaded for TX when called from the oob
+	 * stage.
+	 */
 	return __igb_maybe_stop_tx(tx_ring, size);
 }
 
@@ -6209,7 +6510,21 @@ static int igb_tx_map(struct igb_ring *tx_ring,
 	size = skb_headlen(skb);
 	data_len = skb->data_len;
 
-	dma = dma_map_single(tx_ring->dev, skb->data, size, DMA_TO_DEVICE);
+	if (igb_is_oob_tx(first)) {
+		/*
+		 * An oob-managed packet must have been pre-mapped by
+		 * the sender. We just need to to let the device get
+		 * at that DMA area.
+		 *
+		 * NOTE: oob packets on the TX path should have no
+		 * frags, so we are going to leave the following loop
+		 * early (data_len == 0).
+		 */
+		dma = skb_oob_dma_addr(skb);
+		dma_sync_single_for_device(tx_ring->dev, dma, size, DMA_TO_DEVICE);
+	} else {
+		dma = dma_map_single(tx_ring->dev, skb->data, size, DMA_TO_DEVICE);
+	}
 
 	tx_buffer = first;
 
@@ -6244,6 +6559,12 @@ static int igb_tx_map(struct igb_ring *tx_ring,
 		if (likely(!data_len))
 			break;
 
+		/*
+		 * We should not have frags on the oob TX path (but we
+		 * may have frags on the RX path though).
+		 */
+		WARN_ON_ONCE(igb_running_oob() && dovetail_debug());
+
 		tx_desc->read.cmd_type_len = cpu_to_le32(cmd_type ^ size);
 
 		i++;
@@ -6272,7 +6593,8 @@ static int igb_tx_map(struct igb_ring *tx_ring,
 	/* set the timestamp */
 	first->time_stamp = jiffies;
 
-	skb_tx_timestamp(skb);
+	if (!igb_running_oob())
+		skb_tx_timestamp(skb);
 
 	/* Force memory writes to complete before letting h/w know there
 	 * are new descriptors to fetch.  (Only applicable for weak-ordered
@@ -6318,7 +6640,8 @@ static int igb_tx_map(struct igb_ring *tx_ring,
 		tx_buffer = &tx_ring->tx_buffer_info[i];
 	}
 
-	if (dma_unmap_len(tx_buffer, len))
+	/* tx_buffer == first, don't unmap if oob. */
+	if (!igb_is_oob_tx(tx_buffer) && dma_unmap_len(tx_buffer, len))
 		dma_unmap_single(tx_ring->dev,
 				 dma_unmap_addr(tx_buffer, dma),
 				 dma_unmap_len(tx_buffer, len),
@@ -6474,7 +6797,8 @@ netdev_tx_t igb_xmit_frame_ring(struct sk_buff *skb,
 	first->bytecount = skb->len;
 	first->gso_segs = 1;
 
-	if (unlikely(skb_shinfo(skb)->tx_flags & SKBTX_HW_TSTAMP)) {
+	if (!igb_running_oob() && /* no TX timestamping from the oob stage. */
+		unlikely(skb_shinfo(skb)->tx_flags & SKBTX_HW_TSTAMP)) {
 		struct igb_adapter *adapter = netdev_priv(tx_ring->netdev);
 
 		if (adapter->tstamp_config.tx_type == HWTSTAMP_TX_ON &&
@@ -6492,6 +6816,9 @@ netdev_tx_t igb_xmit_frame_ring(struct sk_buff *skb,
 		}
 	}
 
+	if (igb_net_oob() && skb_is_oob_managed(skb))
+		tx_flags |= IGB_TX_OOB;
+
 	if (skb_vlan_tag_present(skb)) {
 		tx_flags |= IGB_TX_FLAGS_VLAN;
 		tx_flags |= (skb_vlan_tag_get(skb) << IGB_TX_FLAGS_VLAN_SHIFT);
@@ -6864,7 +7191,7 @@ static void igb_perout(struct igb_adapter *adapter, int tsintr_tt)
 	if (pin < 0 || pin >= IGB_N_SDP)
 		return;
 
-	spin_lock(&adapter->tmreg_lock);
+	raw_spin_lock(&adapter->tmreg_lock);
 
 	if (hw->mac.type == e1000_82580 ||
 	    hw->mac.type == e1000_i354 ||
@@ -6929,7 +7256,7 @@ static void igb_perout(struct igb_adapter *adapter, int tsintr_tt)
 	wr32(E1000_TSAUXC, tsauxc);
 	adapter->perout[tsintr_tt].start = ts;
 
-	spin_unlock(&adapter->tmreg_lock);
+	raw_spin_unlock(&adapter->tmreg_lock);
 }
 
 static void igb_extts(struct igb_adapter *adapter, int tsintr_tt)
@@ -6951,9 +7278,9 @@ static void igb_extts(struct igb_adapter *adapter, int tsintr_tt)
 		u64 ns = rd32(auxstmpl);
 
 		ns += ((u64)(rd32(auxstmph) & 0xFF)) << 32;
-		spin_lock_irqsave(&adapter->tmreg_lock, flags);
+		raw_spin_lock_irqsave(&adapter->tmreg_lock, flags);
 		ns = timecounter_cyc2time(&adapter->tc, ns);
-		spin_unlock_irqrestore(&adapter->tmreg_lock, flags);
+		raw_spin_unlock_irqrestore(&adapter->tmreg_lock, flags);
 		ts = ns_to_timespec64(ns);
 	} else {
 		ts.tv_nsec = rd32(auxstmpl);
@@ -6966,13 +7293,38 @@ static void igb_extts(struct igb_adapter *adapter, int tsintr_tt)
 	ptp_clock_event(adapter->ptp_clock, &event);
 }
 
+#ifdef CONFIG_IGB_OOB
+static inline u32 igb_get_tsicr(struct igb_adapter *adapter)
+{
+	struct e1000_hw *hw = &adapter->hw;
+	u32 tsicr;
+
+	if (igb_net_oob() && running_inband()) {
+		tsicr = adapter->cached_tsicr;
+		adapter->cached_tsicr = 0;
+		return tsicr;
+	}
+
+	adapter->cached_tsicr |= rd32(E1000_TSICR);
+
+	return adapter->cached_tsicr;
+}
+#else
+static inline u32 igb_get_tsicr(struct igb_adapter *adapter)
+{
+	struct e1000_hw *hw = &adapter->hw;
+
+	return rd32(E1000_TSICR);
+}
+#endif
+
 static void igb_tsync_interrupt(struct igb_adapter *adapter)
 {
 	const u32 mask = (TSINTR_SYS_WRAP | E1000_TSICR_TXTS |
 			  TSINTR_TT0 | TSINTR_TT1 |
 			  TSINTR_AUTT0 | TSINTR_AUTT1);
 	struct e1000_hw *hw = &adapter->hw;
-	u32 tsicr = rd32(E1000_TSICR);
+	u32 tsicr = igb_get_tsicr(adapter);
 	struct ptp_clock_event event;
 
 	if (hw->mac.type == e1000_82580) {
@@ -6982,6 +7334,9 @@ static void igb_tsync_interrupt(struct igb_adapter *adapter)
 		wr32(E1000_TSICR, tsicr & mask);
 	}
 
+	if (igb_running_oob())
+		return;
+
 	if (tsicr & TSINTR_SYS_WRAP) {
 		event.type = PTP_CLOCK_PPS;
 		if (adapter->ptp_caps.pps)
@@ -7010,39 +7365,43 @@ static irqreturn_t igb_msix_other(int irq, void *data)
 {
 	struct igb_adapter *adapter = data;
 	struct e1000_hw *hw = &adapter->hw;
-	u32 icr = rd32(E1000_ICR);
+	u32 icr = igb_get_icr(adapter);
 	/* reading ICR causes bit 31 of EICR to be cleared */
 
-	if (icr & E1000_ICR_DRSTA)
-		schedule_work(&adapter->reset_task);
+	if (running_inband()) {
+		if (icr & E1000_ICR_DRSTA)
+			schedule_work(&adapter->reset_task);
 
-	if (icr & E1000_ICR_DOUTSYNC) {
-		/* HW is reporting DMA is out of sync */
-		adapter->stats.doosync++;
-		/* The DMA Out of Sync is also indication of a spoof event
-		 * in IOV mode. Check the Wrong VM Behavior register to
-		 * see if it is really a spoof event.
-		 */
-		igb_check_wvbr(adapter);
-	}
+		if (icr & E1000_ICR_DOUTSYNC) {
+			/* HW is reporting DMA is out of sync */
+			adapter->stats.doosync++;
+			/* The DMA Out of Sync is also indication of a spoof event
+			 * in IOV mode. Check the Wrong VM Behavior register to
+			 * see if it is really a spoof event.
+			 */
+			igb_check_wvbr(adapter);
+		}
 
-	/* Check for a mailbox event */
-	if (icr & E1000_ICR_VMMB)
-		igb_msg_task(adapter);
+		/* Check for a mailbox event */
+		if (icr & E1000_ICR_VMMB)
+			igb_msg_task(adapter);
 
-	if (icr & E1000_ICR_LSC) {
-		hw->mac.get_link_status = 1;
-		/* guard against interrupt when we're going down */
-		if (!test_bit(__IGB_DOWN, &adapter->state))
-			mod_timer(&adapter->watchdog_timer, jiffies + 1);
+		if (icr & E1000_ICR_LSC) {
+			hw->mac.get_link_status = 1;
+			/* guard against interrupt when we're going down */
+			if (!test_bit(__IGB_DOWN, &adapter->state))
+				mod_timer(&adapter->watchdog_timer, jiffies + 1);
+		}
 	}
 
 	if (icr & E1000_ICR_TS)
 		igb_tsync_interrupt(adapter);
 
-	wr32(E1000_EIMS, adapter->eims_other);
+	if (!igb_net_oob() || igb_running_oob())
+		wr32(E1000_EIMS, adapter->eims_other);
 
-	return IRQ_HANDLED;
+	return igb_running_oob() && (icr & E1000_ICR_OTHER_FORWARD_MASK)
+		? IRQ_FORWARD : IRQ_HANDLED;
 }
 
 static void igb_write_itr(struct igb_q_vector *q_vector)
@@ -7122,7 +7481,8 @@ static void igb_update_rx_dca(struct igb_adapter *adapter,
 static void igb_update_dca(struct igb_q_vector *q_vector)
 {
 	struct igb_adapter *adapter = q_vector->adapter;
-	int cpu = get_cpu();
+	unsigned long flags;
+	int cpu = hard_get_cpu(flags);
 
 	if (q_vector->cpu == cpu)
 		goto out_no_update;
@@ -7135,7 +7495,7 @@ static void igb_update_dca(struct igb_q_vector *q_vector)
 
 	q_vector->cpu = cpu;
 out_no_update:
-	put_cpu();
+	hard_put_cpu(flags);
 }
 
 static void igb_setup_dca(struct igb_adapter *adapter)
@@ -8092,30 +8452,39 @@ static irqreturn_t igb_intr_msi(int irq, void *data)
 	struct igb_q_vector *q_vector = adapter->q_vector[0];
 	struct e1000_hw *hw = &adapter->hw;
 	/* read ICR disables interrupts using IAM */
-	u32 icr = rd32(E1000_ICR);
+	u32 icr = igb_get_icr(adapter);
 
-	igb_write_itr(q_vector);
+	if (!igb_net_oob() || igb_running_oob())
+		igb_write_itr(q_vector);
 
-	if (icr & E1000_ICR_DRSTA)
-		schedule_work(&adapter->reset_task);
+	if (running_inband()) {
+		if (icr & E1000_ICR_DRSTA)
+			schedule_work(&adapter->reset_task);
 
-	if (icr & E1000_ICR_DOUTSYNC) {
-		/* HW is reporting DMA is out of sync */
-		adapter->stats.doosync++;
-	}
+		if (icr & E1000_ICR_DOUTSYNC) {
+			/* HW is reporting DMA is out of sync */
+			adapter->stats.doosync++;
+		}
 
-	if (icr & (E1000_ICR_RXSEQ | E1000_ICR_LSC)) {
-		hw->mac.get_link_status = 1;
-		if (!test_bit(__IGB_DOWN, &adapter->state))
-			mod_timer(&adapter->watchdog_timer, jiffies + 1);
+		if (icr & (E1000_ICR_RXSEQ | E1000_ICR_LSC)) {
+			hw->mac.get_link_status = 1;
+			if (!test_bit(__IGB_DOWN, &adapter->state))
+				mod_timer(&adapter->watchdog_timer, jiffies + 1);
+		}
 	}
 
 	if (icr & E1000_ICR_TS)
 		igb_tsync_interrupt(adapter);
 
-	napi_schedule(&q_vector->napi);
+	if (!igb_net_oob() || igb_running_oob())
+		napi_schedule(&q_vector->napi);
 
-	return IRQ_HANDLED;
+	/*
+	 * Forwarding means that the current oob handler is going to
+	 * be reentered from the inband stage asap later on.
+	 */
+	return running_oob() && (icr & E1000_ICR_FORWARD_MASK)
+		? IRQ_FORWARD : IRQ_HANDLED;
 }
 
 /**
@@ -8131,37 +8500,44 @@ static irqreturn_t igb_intr(int irq, void *data)
 	/* Interrupt Auto-Mask...upon reading ICR, interrupts are masked.  No
 	 * need for the IMC write
 	 */
-	u32 icr = rd32(E1000_ICR);
+	u32 icr = igb_get_icr(adapter);
 
 	/* IMS will not auto-mask if INT_ASSERTED is not set, and if it is
 	 * not set, then the adapter didn't send an interrupt
 	 */
-	if (!(icr & E1000_ICR_INT_ASSERTED))
+	if (!(icr & E1000_ICR_INT_ASSERTED)) {
+		igb_clear_icr(adapter);
 		return IRQ_NONE;
+	}
 
-	igb_write_itr(q_vector);
+	if (!igb_net_oob() || igb_running_oob())
+		igb_write_itr(q_vector);
 
-	if (icr & E1000_ICR_DRSTA)
-		schedule_work(&adapter->reset_task);
+	if (running_inband()) {
+		if (icr & E1000_ICR_DRSTA)
+			schedule_work(&adapter->reset_task);
 
-	if (icr & E1000_ICR_DOUTSYNC) {
-		/* HW is reporting DMA is out of sync */
-		adapter->stats.doosync++;
-	}
+		if (icr & E1000_ICR_DOUTSYNC) {
+			/* HW is reporting DMA is out of sync */
+			adapter->stats.doosync++;
+		}
 
-	if (icr & (E1000_ICR_RXSEQ | E1000_ICR_LSC)) {
-		hw->mac.get_link_status = 1;
-		/* guard against interrupt when we're going down */
-		if (!test_bit(__IGB_DOWN, &adapter->state))
-			mod_timer(&adapter->watchdog_timer, jiffies + 1);
+		if (icr & (E1000_ICR_RXSEQ | E1000_ICR_LSC)) {
+			hw->mac.get_link_status = 1;
+			/* guard against interrupt when we're going down */
+			if (!test_bit(__IGB_DOWN, &adapter->state))
+				mod_timer(&adapter->watchdog_timer, jiffies + 1);
+		}
 	}
 
 	if (icr & E1000_ICR_TS)
 		igb_tsync_interrupt(adapter);
 
-	napi_schedule(&q_vector->napi);
+	if (!igb_net_oob() || igb_running_oob())
+		napi_schedule(&q_vector->napi);
 
-	return IRQ_HANDLED;
+	return running_oob() && (icr & E1000_ICR_FORWARD_MASK)
+		? IRQ_FORWARD : IRQ_HANDLED;
 }
 
 static void igb_ring_irq_enable(struct igb_q_vector *q_vector)
@@ -8277,11 +8653,9 @@ static bool igb_clean_tx_irq(struct igb_q_vector *q_vector, int napi_budget)
 		else
 			xdp_return_frame(tx_buffer->xdpf);
 
-		/* unmap skb header data */
-		dma_unmap_single(tx_ring->dev,
-				 dma_unmap_addr(tx_buffer, dma),
-				 dma_unmap_len(tx_buffer, len),
-				 DMA_TO_DEVICE);
+		/* unmap skb header data unless oob-managed */
+		if (!igb_is_oob_tx(tx_buffer))
+			igb_unmap_buffer(tx_ring, tx_buffer);
 
 		/* clear tx_buffer data */
 		dma_unmap_len_set(tx_buffer, len, 0);
@@ -8299,10 +8673,8 @@ static bool igb_clean_tx_irq(struct igb_q_vector *q_vector, int napi_budget)
 
 			/* unmap any remaining paged data */
 			if (dma_unmap_len(tx_buffer, len)) {
-				dma_unmap_page(tx_ring->dev,
-					       dma_unmap_addr(tx_buffer, dma),
-					       dma_unmap_len(tx_buffer, len),
-					       DMA_TO_DEVICE);
+				if (!igb_is_oob_tx(tx_buffer))
+					igb_unmap_buffer(tx_ring, tx_buffer);
 				dma_unmap_len_set(tx_buffer, len, 0);
 			}
 		}
@@ -8405,7 +8777,8 @@ static bool igb_clean_tx_irq(struct igb_q_vector *q_vector, int napi_budget)
  *  @rx_ring: rx descriptor ring to store buffers on
  *  @old_buff: donor buffer to have page reused
  *
- *  Synchronizes page for reuse by the adapter
+ *  Synchronizes page for reuse by the adapter.
+ *  CAUTION: We must _never_ reuse oob pages.
  **/
 static void igb_reuse_rx_page(struct igb_ring *rx_ring,
 			      struct igb_rx_buffer *old_buff)
@@ -8429,14 +8802,15 @@ static void igb_reuse_rx_page(struct igb_ring *rx_ring,
 	new_buff->pagecnt_bias	= old_buff->pagecnt_bias;
 }
 
-static bool igb_can_reuse_rx_page(struct igb_rx_buffer *rx_buffer,
-				  int rx_buf_pgcnt)
+static bool igb_can_reuse_rx_page(struct igb_ring *rx_ring,
+				struct igb_rx_buffer *rx_buffer,
+				int rx_buf_pgcnt)
 {
 	unsigned int pagecnt_bias = rx_buffer->pagecnt_bias;
 	struct page *page = rx_buffer->page;
 
-	/* avoid re-using remote and pfmemalloc pages */
-	if (!dev_page_is_reusable(page))
+	/* avoid re-using oob, remote and pfmemalloc pages */
+	if (igb_is_oob_page(rx_ring, page) || !dev_page_is_reusable(page))
 		return false;
 
 #if (PAGE_SIZE < 8192)
@@ -8508,6 +8882,9 @@ static struct sk_buff *igb_construct_skb(struct igb_ring *rx_ring,
 	unsigned int headlen;
 	struct sk_buff *skb;
 
+	/* This is legacy RX: not supported in oob mode. */
+	WARN_ON_ONCE(igb_running_oob() && dovetail_debug());
+
 	/* prefetch first cache line of first page */
 	net_prefetch(xdp->data);
 
@@ -8568,6 +8945,9 @@ static struct sk_buff *igb_build_skb(struct igb_ring *rx_ring,
 	if (unlikely(!skb))
 		return NULL;
 
+	if (igb_is_oob_page(rx_ring, rx_buffer->page))
+		skb_mark_for_recycle(skb);
+
 	/* update pointers within the skb to store the data */
 	skb_reserve(skb, xdp->data - xdp->data_hard_start);
 	__skb_put(skb, xdp->data_end - xdp->data);
@@ -8852,19 +9232,15 @@ static struct igb_rx_buffer *igb_get_rx_buffer(struct igb_ring *rx_ring,
 static void igb_put_rx_buffer(struct igb_ring *rx_ring,
 			      struct igb_rx_buffer *rx_buffer, int rx_buf_pgcnt)
 {
-	if (igb_can_reuse_rx_page(rx_buffer, rx_buf_pgcnt)) {
+	if (igb_can_reuse_rx_page(rx_ring, rx_buffer, rx_buf_pgcnt))
 		/* hand second half of page back to the ring */
 		igb_reuse_rx_page(rx_ring, rx_buffer);
-	} else {
+	else if (!igb_is_oob_page(rx_ring, rx_buffer->page))
 		/* We are not reusing the buffer so unmap it and free
-		 * any references we are holding to it
-		 */
-		dma_unmap_page_attrs(rx_ring->dev, rx_buffer->dma,
-				     igb_rx_pg_size(rx_ring), DMA_FROM_DEVICE,
-				     IGB_RX_DMA_ATTR);
-		__page_frag_cache_drain(rx_buffer->page,
-					rx_buffer->pagecnt_bias);
-	}
+		 * any references we are holding to the page. oob
+		 * pages never reused, always released by the skb
+		 * consumer. */
+		igb_flush_page(rx_ring, rx_buffer);
 
 	/* clear contents of rx_buffer */
 	rx_buffer->page = NULL;
@@ -8940,7 +9316,8 @@ static int igb_clean_rx_irq(struct igb_q_vector *q_vector, const int budget)
 			/* At larger PAGE_SIZE, frame_sz depend on len size */
 			xdp.frame_sz = igb_rx_frame_truesize(rx_ring, size);
 #endif
-			skb = igb_run_xdp(adapter, rx_ring, &xdp);
+			if (!igb_running_oob()) /* Bypass XDP layer if oob */
+				skb = igb_run_xdp(adapter, rx_ring, &xdp);
 		}
 
 		if (IS_ERR(skb)) {
@@ -9037,32 +9414,48 @@ static bool igb_alloc_mapped_page(struct igb_ring *rx_ring,
 		return true;
 
 	/* alloc new page for storage */
-	page = dev_alloc_pages(igb_rx_pg_order(rx_ring));
+#ifdef CONFIG_IGB_OOB
+	if (igb_running_oob())
+		page = page_pool_dev_alloc_pages(rx_ring->rx_oob_pool);
+	else
+#endif
+		page = dev_alloc_pages(igb_rx_pg_order(rx_ring));
+
 	if (unlikely(!page)) {
+		WARN_ON_ONCE(1);
 		rx_ring->rx_stats.alloc_failed++;
 		return false;
 	}
 
-	/* map page for use */
-	dma = dma_map_page_attrs(rx_ring->dev, page, 0,
-				 igb_rx_pg_size(rx_ring),
-				 DMA_FROM_DEVICE,
-				 IGB_RX_DMA_ATTR);
+	/* unless coming from the oob pool which delivers pre-mapped
+	 * memory, map page */
+	if (!igb_is_oob_page(rx_ring, page)) {
+		dma = dma_map_page_attrs(rx_ring->dev, page, 0,
+					igb_rx_pg_size(rx_ring),
+					DMA_FROM_DEVICE,
+					IGB_RX_DMA_ATTR);
 
-	/* if mapping failed free memory back to system since
-	 * there isn't much point in holding memory we can't use
-	 */
-	if (dma_mapping_error(rx_ring->dev, dma)) {
-		__free_pages(page, igb_rx_pg_order(rx_ring));
+		/* if mapping failed free memory back to system since
+		 * there isn't much point in holding memory we can't use
+		 */
+		if (dma_mapping_error(rx_ring->dev, dma)) {
+			__free_pages(page, igb_rx_pg_order(rx_ring));
 
-		rx_ring->rx_stats.alloc_failed++;
-		return false;
+			rx_ring->rx_stats.alloc_failed++;
+			return false;
+		}
+		page_ref_add(page, USHRT_MAX - 1);
+	} else {
+		/* oob pages are never reused, always released by the
+		 * skb consumer, so the initial refcount value for
+		 * these is 1, and we actually ignore the bias
+		 * count. */
+		dma = page_pool_get_dma_addr(page);
 	}
 
 	bi->dma = dma;
 	bi->page = page;
 	bi->page_offset = igb_rx_offset(rx_ring);
-	page_ref_add(page, USHRT_MAX - 1);
 	bi->pagecnt_bias = USHRT_MAX;
 
 	return true;
diff --git a/drivers/net/ethernet/intel/igb/igb_ptp.c b/drivers/net/ethernet/intel/igb/igb_ptp.c
index f94570556120..93625401a87c 100644
--- a/drivers/net/ethernet/intel/igb/igb_ptp.c
+++ b/drivers/net/ethernet/intel/igb/igb_ptp.c
@@ -173,9 +173,9 @@ static void igb_ptp_systim_to_hwtstamp(struct igb_adapter *adapter,
 	case e1000_82580:
 	case e1000_i354:
 	case e1000_i350:
-		spin_lock_irqsave(&adapter->tmreg_lock, flags);
+		raw_spin_lock_irqsave(&adapter->tmreg_lock, flags);
 		ns = timecounter_cyc2time(&adapter->tc, systim);
-		spin_unlock_irqrestore(&adapter->tmreg_lock, flags);
+		raw_spin_unlock_irqrestore(&adapter->tmreg_lock, flags);
 
 		hwtstamps->hwtstamp = ns_to_ktime(ns);
 		break;
@@ -231,9 +231,9 @@ static int igb_ptp_adjtime_82576(struct ptp_clock_info *ptp, s64 delta)
 					       ptp_caps);
 	unsigned long flags;
 
-	spin_lock_irqsave(&igb->tmreg_lock, flags);
+	raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 	timecounter_adjtime(&igb->tc, delta);
-	spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+	raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 
 	return 0;
 }
@@ -245,13 +245,13 @@ static int igb_ptp_adjtime_i210(struct ptp_clock_info *ptp, s64 delta)
 	unsigned long flags;
 	struct timespec64 now, then = ns_to_timespec64(delta);
 
-	spin_lock_irqsave(&igb->tmreg_lock, flags);
+	raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 
 	igb_ptp_read_i210(igb, &now);
 	now = timespec64_add(now, then);
 	igb_ptp_write_i210(igb, (const struct timespec64 *)&now);
 
-	spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+	raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 
 	return 0;
 }
@@ -267,7 +267,7 @@ static int igb_ptp_gettimex_82576(struct ptp_clock_info *ptp,
 	u32 lo, hi;
 	u64 ns;
 
-	spin_lock_irqsave(&igb->tmreg_lock, flags);
+	raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 
 	ptp_read_system_prets(sts);
 	lo = rd32(E1000_SYSTIML);
@@ -276,7 +276,7 @@ static int igb_ptp_gettimex_82576(struct ptp_clock_info *ptp,
 
 	ns = timecounter_cyc2time(&igb->tc, ((u64)hi << 32) | lo);
 
-	spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+	raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 
 	*ts = ns_to_timespec64(ns);
 
@@ -294,7 +294,7 @@ static int igb_ptp_gettimex_82580(struct ptp_clock_info *ptp,
 	u32 lo, hi;
 	u64 ns;
 
-	spin_lock_irqsave(&igb->tmreg_lock, flags);
+	raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 
 	ptp_read_system_prets(sts);
 	rd32(E1000_SYSTIMR);
@@ -304,7 +304,7 @@ static int igb_ptp_gettimex_82580(struct ptp_clock_info *ptp,
 
 	ns = timecounter_cyc2time(&igb->tc, ((u64)hi << 32) | lo);
 
-	spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+	raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 
 	*ts = ns_to_timespec64(ns);
 
@@ -320,7 +320,7 @@ static int igb_ptp_gettimex_i210(struct ptp_clock_info *ptp,
 	struct e1000_hw *hw = &igb->hw;
 	unsigned long flags;
 
-	spin_lock_irqsave(&igb->tmreg_lock, flags);
+	raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 
 	ptp_read_system_prets(sts);
 	rd32(E1000_SYSTIMR);
@@ -328,7 +328,7 @@ static int igb_ptp_gettimex_i210(struct ptp_clock_info *ptp,
 	ts->tv_nsec = rd32(E1000_SYSTIML);
 	ts->tv_sec = rd32(E1000_SYSTIMH);
 
-	spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+	raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 
 	return 0;
 }
@@ -343,11 +343,11 @@ static int igb_ptp_settime_82576(struct ptp_clock_info *ptp,
 
 	ns = timespec64_to_ns(ts);
 
-	spin_lock_irqsave(&igb->tmreg_lock, flags);
+	raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 
 	timecounter_init(&igb->tc, &igb->cc, ns);
 
-	spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+	raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 
 	return 0;
 }
@@ -359,11 +359,11 @@ static int igb_ptp_settime_i210(struct ptp_clock_info *ptp,
 					       ptp_caps);
 	unsigned long flags;
 
-	spin_lock_irqsave(&igb->tmreg_lock, flags);
+	raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 
 	igb_ptp_write_i210(igb, ts);
 
-	spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+	raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 
 	return 0;
 }
@@ -522,7 +522,7 @@ static int igb_ptp_feature_enable_82580(struct ptp_clock_info *ptp,
 			tsauxc_mask = TSAUXC_EN_TS0;
 			tsim_mask = TSINTR_AUTT0;
 		}
-		spin_lock_irqsave(&igb->tmreg_lock, flags);
+		raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 		tsauxc = rd32(E1000_TSAUXC);
 		tsim = rd32(E1000_TSIM);
 		if (on) {
@@ -535,7 +535,7 @@ static int igb_ptp_feature_enable_82580(struct ptp_clock_info *ptp,
 		}
 		wr32(E1000_TSAUXC, tsauxc);
 		wr32(E1000_TSIM, tsim);
-		spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+		raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 		return 0;
 
 	case PTP_CLK_REQ_PEROUT:
@@ -567,7 +567,7 @@ static int igb_ptp_feature_enable_82580(struct ptp_clock_info *ptp,
 			trgttiml = E1000_TRGTTIML0;
 			trgttimh = E1000_TRGTTIMH0;
 		}
-		spin_lock_irqsave(&igb->tmreg_lock, flags);
+		raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 		tsauxc = rd32(E1000_TSAUXC);
 		tsim = rd32(E1000_TSIM);
 		if (rq->perout.index == 1) {
@@ -628,7 +628,7 @@ static int igb_ptp_feature_enable_82580(struct ptp_clock_info *ptp,
 		}
 		wr32(E1000_TSAUXC, tsauxc);
 		wr32(E1000_TSIM, tsim);
-		spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+		raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 		return 0;
 
 	case PTP_CLK_REQ_PPS:
@@ -678,7 +678,7 @@ static int igb_ptp_feature_enable_i210(struct ptp_clock_info *ptp,
 			tsauxc_mask = TSAUXC_EN_TS0;
 			tsim_mask = TSINTR_AUTT0;
 		}
-		spin_lock_irqsave(&igb->tmreg_lock, flags);
+		raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 		tsauxc = rd32(E1000_TSAUXC);
 		tsim = rd32(E1000_TSIM);
 		if (on) {
@@ -691,7 +691,7 @@ static int igb_ptp_feature_enable_i210(struct ptp_clock_info *ptp,
 		}
 		wr32(E1000_TSAUXC, tsauxc);
 		wr32(E1000_TSIM, tsim);
-		spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+		raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 		return 0;
 
 	case PTP_CLK_REQ_PEROUT:
@@ -739,7 +739,7 @@ static int igb_ptp_feature_enable_i210(struct ptp_clock_info *ptp,
 			trgttimh = E1000_TRGTTIMH0;
 			freqout = E1000_FREQOUT0;
 		}
-		spin_lock_irqsave(&igb->tmreg_lock, flags);
+		raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 		tsauxc = rd32(E1000_TSAUXC);
 		tsim = rd32(E1000_TSIM);
 		if (rq->perout.index == 1) {
@@ -765,11 +765,11 @@ static int igb_ptp_feature_enable_i210(struct ptp_clock_info *ptp,
 		}
 		wr32(E1000_TSAUXC, tsauxc);
 		wr32(E1000_TSIM, tsim);
-		spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+		raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 		return 0;
 
 	case PTP_CLK_REQ_PPS:
-		spin_lock_irqsave(&igb->tmreg_lock, flags);
+		raw_spin_lock_irqsave(&igb->tmreg_lock, flags);
 		tsim = rd32(E1000_TSIM);
 		if (on)
 			tsim |= TSINTR_SYS_WRAP;
@@ -777,7 +777,7 @@ static int igb_ptp_feature_enable_i210(struct ptp_clock_info *ptp,
 			tsim &= ~TSINTR_SYS_WRAP;
 		igb->pps_sys_wrap_on = !!on;
 		wr32(E1000_TSIM, tsim);
-		spin_unlock_irqrestore(&igb->tmreg_lock, flags);
+		raw_spin_unlock_irqrestore(&igb->tmreg_lock, flags);
 		return 0;
 	}
 
@@ -1396,7 +1396,7 @@ void igb_ptp_init(struct igb_adapter *adapter)
 			 adapter->netdev->name);
 		adapter->ptp_flags |= IGB_PTP_ENABLED;
 
-		spin_lock_init(&adapter->tmreg_lock);
+		raw_spin_lock_init(&adapter->tmreg_lock);
 		INIT_WORK(&adapter->ptp_tx_work, igb_ptp_tx_work);
 
 		if (adapter->ptp_flags & IGB_PTP_OVERFLOW_CHECK)
@@ -1482,7 +1482,7 @@ void igb_ptp_reset(struct igb_adapter *adapter)
 	/* reset the tstamp_config */
 	igb_ptp_set_timestamp_mode(adapter, &adapter->tstamp_config);
 
-	spin_lock_irqsave(&adapter->tmreg_lock, flags);
+	raw_spin_lock_irqsave(&adapter->tmreg_lock, flags);
 
 	switch (adapter->hw.mac.type) {
 	case e1000_82576:
@@ -1516,7 +1516,7 @@ void igb_ptp_reset(struct igb_adapter *adapter)
 				 ktime_to_ns(ktime_get_real()));
 	}
 out:
-	spin_unlock_irqrestore(&adapter->tmreg_lock, flags);
+	raw_spin_unlock_irqrestore(&adapter->tmreg_lock, flags);
 
 	wrfl();
 
-- 
2.30.2

